uaclient/files/data_types.py

```diff
@@ -0,0 +1,68 @@
+import json
+from enum import Enum
+from typing import Any, Dict, Generic, Optional, Type, TypeVar
+
+import yaml
+
+from uaclient import exceptions
+from uaclient.data_types import DataObject
+from uaclient.files.files import UAFile
+
+
+class DataObjectFileFormat(Enum):
+    JSON = "json"
+    YAML = "yaml"
+
+
+DOFType = TypeVar("DOFType", bound=DataObject)
+
+
+class DataObjectFile(Generic[DOFType]):
+    def __init__(
+        self,
+        data_object_cls: Type[DOFType],
+        ua_file: UAFile,
+        file_format: DataObjectFileFormat = DataObjectFileFormat.JSON,
+    ):
+        self.data_object_cls = data_object_cls
+        self.ua_file = ua_file
+        self.file_format = file_format
+
+    def preprocess_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        return data
+
+    def read(self) -> Optional[DOFType]:
+        raw_data = self.ua_file.read()
+        if raw_data is None:
+            return None
+
+        parsed_data = None
+        if self.file_format == DataObjectFileFormat.JSON:
+            try:
+                parsed_data = json.loads(raw_data)
+            except json.JSONDecodeError:
+                raise exceptions.InvalidFileFormatError(
+                    self.ua_file.path, "json"
+                )
+        elif self.file_format == DataObjectFileFormat.YAML:
+            try:
+                parsed_data = yaml.safe_load(raw_data)
+            except yaml.parser.ParserError:
+                raise exceptions.InvalidFileFormatError(
+                    self.ua_file.path, "yaml"
+                )
+
+        if parsed_data is None:
+            return None
+
+        parsed_data = self.preprocess_data(parsed_data)
```

Comment:
This preprocess hook will be useful for any other migrations we may want to do in the future. +1